ValidateOtp/handler.ts

```diff
@@ -139,37 +151,58 @@ export function ValidateOtpHandler(
       // tslint:disable: no-useless-cast
       payload.otp_code.toString() as NonEmptyString
     );
-    return retrieveOtp(redisClient, payload.otp_code)
-      .mapLeft<IResponseErrorInternal | IResponseErrorNotFound>(_ =>
-        errorLogMapping(_, ResponseErrorInternal("Cannot validate OTP Code"))
-      )
-      .chain<OtpValidationResponse>(maybeOtpResponseAndFiscalCode =>
-        maybeOtpResponseAndFiscalCode.foldL(
-          () =>
-            fromLeft(
-              ResponseErrorNotFound("Not Found", "OTP Not Found or invalid")
-            ),
-          otpResponseAndFiscalCode =>
-            payload.invalidate_otp
-              ? invalidateOtp(
-                  redisClient,
-                  payload.otp_code,
-                  otpResponseAndFiscalCode.fiscalCode
-                ).bimap(
-                  _ =>
-                    errorLogMapping(
-                      _,
-                      ResponseErrorInternal("Cannot invalidate OTP")
+
+    const result = pipe(
+      retrieveOtp(redisClient, payload.otp_code),
+      TE.mapLeft(e =>
+        errorLogMapping(e, ResponseErrorInternal("Cannot validate OTP Code"))
+      ),
+      TE.chain<
+        IResponseErrorInternal | IResponseErrorNotFound,
+        O.Option<OtpResponseAndFiscalCode>,
+        OtpValidationResponse
+      >(
+        flow(
+          O.fold(
+            () =>
+              TE.left(
+                ResponseErrorNotFound("Not Found", "OTP Not Found or invalid")
+              ),
+            otpResponseAndFiscalCode =>
+              payload.invalidate_otp
+                ? pipe(
+                    invalidateOtp(
+                      redisClient,
+                      payload.otp_code,
+                      otpResponseAndFiscalCode.fiscalCode
                     ),
-                  () => ({
-                    expires_at: new Date()
-                  })
-                )
-              : taskEither.of(otpResponseAndFiscalCode.otpResponse)
+                    TE.bimap(
+                      e =>
+                        errorLogMapping(
+                          e,
+                          ResponseErrorInternal("Cannot invalidate OTP")
+                        ),
+                      () => ({
+                        expires_at: new Date()
+                      })
+                    )
+                  )
+                : TE.of(otpResponseAndFiscalCode.otpResponse)
+          )
         )
+      ),
+      TE.foldW<
+        IResponseErrorInternal | IResponseErrorNotFound,
+        ResponseTypes,
+        OtpValidationResponse,
+        ResponseTypes
+      >(
+        e => T.of(e),
+        s => T.of(ResponseSuccessJson(s))
```

Comment:
why folding here instead of using `TE.orElse` ?

Reply:
This is just a "translation" of old logic. I'll investigate to improve it.

Reply:
Changed to toUnion by merging your fp-ts2 porting. I kept the best of both.